### ISSUE TITLE
Close file handles

### ISSUE THREAD

**Issue by hackartisan** [pulibrary/figgy#4529]
They will appear to be on still but 0/5 will be working, while enqueued jobs build up.

The first time I encountered this was April 16th.

Is there some change in our dependencies that happened at this time? Is something going wrong between sidekiq and redis?

They seem to run fine for at least a few days once restarted.

## Update 1/14/22

We're noticing that workers are surviving for weeks at a time with no problems now. Our Valkyrie update improved the situation, and now it's possible to close file handles.

To close this ticket, all instances of workers which open TempFiles or Valkyrie::Files need to be closed via `#close` explicitly.

**Comment by tpendragon** [comment]
TL;DR: We need to wait for one of these to break to get a thread backtrace to figure out if we have a thread leak.

https://app.honeybadger.io/projects/53391/faults/35108866/01F65P9QDDF1EEW4EEEB18WG9B is the error that's killing these. As soon as this happens that worker's dead and can no longer pick up jobs.

According to https://github.com/mperham/sidekiq/issues/2862 this is likely because there's a thread leak in our code somewhere - something's spawning a thread and not getting rid of itself, we'll have to make sure that thread-spawner's not run in the Sidekiq process. Maybe Bunny? That'd be my guess.

We can find out by waiting until a worker dies, leaving it, and doing the following:

In one pane:

`sudo tail -f /var/log/syslog | grep sidekiq`
`ps aux | grep sidekiq` (get sidekiq Process ID)
`sudo kill -TTIN sidekiq_pid`

Then copy the backtrace of all the threads into this issue so we can diagnose them. The TTIN signal will dump the backtrace, as per https://github.com/mperham/sidekiq/wiki/Signals#ttin

**Comment by hackartisan** [comment]
Okay I tried this investigation since lib-proc7 had a PdfOcrJob running since 6 days ago. thread count was 19. Not sure what "pretty high" means but I did see someone on the internet mention 89 threads so it's definitely not as high as that.

Out of curiosity, I checked thread counts on the other boxes. most were in the 16-19 range. Two of them (proc9 and proc11) had a thread count of 1. I restarted workers on a couple of them and they came up with 12 threads.

**Comment by tpendragon** [comment]
Apparently Valkyrie can register a finalizer to close the IO stream if it's garbage collected.

**Comment by tpendragon** [comment]
So I think Valkyrie needs to:

* Provide access to #close 
* Not open a file descriptor until you're trying to read the content of a file.

**Comment by eliotjordan** [comment]
@tpendragon Can we close this?
https://github.com/pulibrary/figgy/pull/5381
